### PR TITLE
Schema contracts migration: Arrow-native pipelines on alias identity

### PR DESCRIPTION
**Pull request by kirill-andr**
## Summary
- Plumb Arrow `RecordBatch` end-to-end through source → engine → destination via a shared `SchemaContract`, replacing row-by-row coercion.
- Move pipeline identity onto connection aliases and resolve connector kind / contract schemas at runtime instead of off the endpoint document.
- Add JSON-Schema validation in `src/config/schema_validator.py` and dispatch API destination writes on mode-keyed `operations.write.<mode>`.
- Tidy: drop unused `connector_utils`, untrack stale `__pycache__` bytecode.

## Test plan
- [ ] `poetry run pytest` (full suite)
- [ ] Run a representative pipeline end-to-end in Docker (`PIPELINE_ID=... docker compose run --rm source_engine`) and confirm clean source→destination flow
- [ ] Spot-check API destination upsert/insert/replace dispatch
- [ ] Spot-check database destination DDL + Arrow batch cast on a fresh table

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/engine/pipeline.py [hunk 1]
```diff
@@ -234,179 +115,189 @@ def get_endpoint_config(endpoint_ref: Any) -> Dict[str, Any]:
...
+        target_path = list(target_path_raw)
+    else:
+        target_path = []
+
+    # Pass the contract-declared fields through unchanged. The transformer
+    # treats unknown ``type`` values as passthrough (no JSON coercion),
+    # which is the right behavior for Arrow-typed values headed to the
+    # destination's Arrow-based schema contract.
+    target = dict(raw_target)
+    target["path"] = target_path
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply src/engine/pipeline.py:282]: Reviewed against the current code — the transformer no longer reads `target.get("type")`. `AssignmentTransformer.transform_record` (now `DataTransformer.transform_record` in `src/engine/data_transformer.py:53-123`) only consults `target.path` and `target.nullable`; the typed Arrow batch is built by `build_output_schema` (`data_transformer.py:702-732`) which reads `target.arrow_type` and inherits the input type for `op:"get"` passthroughs. There is no string-coercion fallback to preserve. Closing as not applicable.

File: src/config/endpoint_resolver.py [hunk 2]
```diff
@@ -1,81 +1,119 @@
...
-        / f"{parsed.endpoint}.json"
-    )
+    if parsed.scope == "connector":
+        connector_alias = lookup.connector_alias_for(parsed.connection_id)
+        file_path = (
+            paths["connectors"]
+            / connector_alias
+            / "definition"
+            / "endpoints"
+            / f"{parsed.alias}.json"
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply src/config/endpoint_resolver.py]: Fixed in f455643. Added a second candidate `connectors/connector-{alias}/definition/endpoints/{name}.json` matching `load_connector_definition`'s two-layout lookup; behavior preserved for the `{alias}` form already used by the green pipelines.

File: src/destination/connectors/api.py [hunk 3]
```diff
@@ -4,11 +4,14 @@
 rate limiting, retries, and different batch modes.
 """
 
+import asyncio
 import json
 import logging
-from typing import Any, Dict, List, Optional
+from dataclasses import dataclass
+from typing import Any, Dict, List, Mapping, Optional
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply src/destination/connectors/api.py]: Fixed in f455643: removed unused `Optional` from the typing import.

File: src/source/connectors/api.py [hunk 4]
```diff
@@ -45,758 +57,623 @@ async def connect(self, runtime: ConnectionRuntime):
...
+                isinstance(stored_value, str)
+                and str(record_value).isdigit()
+                and stored_value.isdigit()
+            ):
+                if int(record_value) > int(stored_value):
+                    return True
+                if int(record_value) < int(stored_value):
+                    return False
+                continue
+        except (ValueError, TypeError):
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply src/source/connectors/api.py:726]: Fixed in f455643: added an inline comment explaining the fallback (Unicode digit edge case where `str.isdigit()` is True but `int()` raises; falls through to lexicographic compare).

File: tests/unit/destination/handlers/test_database_handler_endpoint_refs.py [hunk 5]
```diff
@@ -178,20 +154,117 @@ async def _not_committed(*_args, **_kwargs):
...
+        async def _fake_run_sync(_callable, *_args, **_kwargs):
+            nonlocal in_flight, max_concurrent
+            in_flight += 1
+            max_concurrent = max(max_concurrent, in_flight)
+            # Yield control to give a parallel coroutine a chance to enter
+            # this critical section if the lock isn't holding.
+            await asyncio.sleep(0.01)
+            in_flight -= 1
+
+        async def _fake_begin():
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply tests/unit/destination/handlers/test_database_handler_endpoint_refs.py]: Fixed in f455643: removed the unused `_fake_begin` block; the test exclusively uses `_begin_ctx` (the `@asynccontextmanager`-decorated context).

File: tests/unit/test_review_gap_coverage.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+
+        # Seed stale state from a hypothetical prior failed run.
+        client._task_failure = RuntimeError("prev run")
+        client._peer_closed_stream = True
+
+        # Make start_stream's schema-ack wait complete immediately via a
+        # task that fails — we only care about the pre-send reset.
+        sent: list[bool] = []
+
+        async def _fake_create_task(coro, name=None):
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply tests/unit/test_review_gap_coverage.py]: Fixed in f455643: removed the unused `_fake_create_task` and `sent` list; the test uses `_capture_create_task` instead.

File: tests/unit/destination/test_schema_contract.py [hunk 7]
```diff
@@ -37,86 +39,163 @@ def type_mapper() -> TypeMapper:
...
+    """Source-side: building an Arrow batch from dicts."""
+
+    def test_from_pylist_with_schema(self, type_mapper):
+        schema = {
+            "columns": [
+                {"name": "id", "native_type": "BIGINT", "nullable": False},
+                {"name": "name", "native_type": "VARCHAR(100)", "nullable": True},
+            ]
+        }
+        contract = SchemaContract(schema, type_mapper=type_mapper)
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply tests/unit/destination/test_schema_contract.py]: Fixed in f455643: rewrote `tests/unit/destination/test_schema_contract.py` against the current contract API. The contract no longer accepts a `type_mapper=` kwarg (destination side reads `arrow_type` directly off the endpoint document); the new tests declare every column via `arrow_type` and exercise the strptime, Decimal-from-float, and validation paths.

File: src/destination/connectors/api.py [hunk 8]
```diff
@@ -6,9 +6,11 @@
 
 import json
 import logging
-from typing import Any, Dict, List, Optional
+from dataclasses import dataclass
+from typing import Any, Dict, List, Mapping, Optional
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply src/destination/connectors/api.py]: Resolved in f455643 — same fix as the later duplicate of this finding.

File: src/source/connectors/api.py [hunk 9]
```diff
@@ -45,758 +57,619 @@ async def connect(self, runtime: ConnectionRuntime):
...
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply src/source/connectors/api.py:726]: Resolved in f455643 — same fix as the later duplicate of this finding.

File: src/source/connectors/api.py [hunk 10]
```diff
@@ -45,758 +57,621 @@ async def connect(self, runtime: ConnectionRuntime):
...
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply src/source/connectors/api.py:726]: Resolved in f455643 — same fix as the later duplicate of this finding.

File: tests/unit/destination/handlers/test_database_handler_endpoint_refs.py [hunk 11]
```diff
@@ -178,20 +154,99 @@ async def _not_committed(*_args, **_kwargs):
...
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply tests/unit/destination/handlers/test_database_handler_endpoint_refs.py]: Resolved in f455643 — same fix as the later duplicate of this finding.

File: tests/unit/destination/test_schema_contract.py [hunk 12]
```diff
@@ -37,86 +39,163 @@ def type_mapper() -> TypeMapper:
...
-                {"name": "name", "type": "VARCHAR(100)", "nullable": True},
-                {"name": "created", "type": "TIMESTAMP", "nullable": True},
+                {"name": "id", "native_type": "BIGINT", "nullable": False},
+                {"name": "name", "native_type": "VARCHAR(100)", "nullable": True},
+                {"name": "created", "native_type": "TIMESTAMP", "nullable": True},
             ]
         }
 
-        contract = DestinationSchemaContract(schema, type_mapper=type_mapper)
+        contract = SchemaContract(schema, type_mapper=type_mapper)
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply tests/unit/destination/test_schema_contract.py]: Resolved in f455643 — same fix as the later duplicate of this finding.

File: tests/unit/destination/test_schema_contract.py [hunk 13]
```diff
@@ -37,86 +39,163 @@ def type_mapper() -> TypeMapper:
...
+    def test_cast_arrow_batch_basic(self, type_mapper):
         schema = {
             "columns": [
-                {"name": "id", "type": "BIGINT", "nullable": False},
-                {"name": "name", "type": "VARCHAR(100)", "nullable": True},
+                {"name": "id", "native_type": "BIGINT", "nullable": False},
+                {"name": "name", "native_type": "VARCHAR(100)", "nullable": True},
             ]
         }
+        contract = SchemaContract(schema, type_mapper=type_mapper)
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply tests/unit/destination/test_schema_contract.py]: Resolved in f455643 — same fix as the later duplicate of this finding.

File: tests/unit/destination/test_schema_contract.py [hunk 14]
```diff
@@ -37,86 +39,163 @@ def type_mapper() -> TypeMapper:
...
+    def test_cast_arrow_batch_with_type_coercion(self, type_mapper):
         schema = {
             "columns": [
-                {"name": "id", "type": "BIGINT", "nullable": False},
-                {"name": "value", "type": "FLOAT", "nullable": True},
+                {"name": "id", "native_type": "BIGINT", "nullable": False},
+                {"name": "value", "native_type": "FLOAT", "nullable": True},
             ]
         }
+        contract = SchemaContract(schema, type_mapper=type_mapper)
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply tests/unit/destination/test_schema_contract.py]: Resolved in f455643 — same fix as the later duplicate of this finding.

File: tests/unit/destination/test_schema_contract.py [hunk 15]
```diff
@@ -37,86 +39,163 @@ def type_mapper() -> TypeMapper:
...
         assert result[0]["id"] == 1
         assert isinstance(result[0]["value"], float)
 
-    def test_cast_batch_empty(self, type_mapper):
-        schema = {"columns": [{"name": "id", "type": "BIGINT", "nullable": False}]}
-        contract = DestinationSchemaContract(schema, type_mapper=type_mapper)
-        assert contract.prepare_records([]) == []
+    def test_cast_arrow_batch_empty(self, type_mapper):
+        schema = {"columns": [{"name": "id", "native_type": "BIGINT", "nullable": False}]}
+        contract = SchemaContract(schema, type_mapper=type_mapper)
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply tests/unit/destination/test_schema_contract.py]: Resolved in f455643 — same fix as the later duplicate of this finding.

File: tests/unit/destination/test_schema_contract.py [hunk 16]
```diff
@@ -37,86 +39,163 @@ def type_mapper() -> TypeMapper:
...
         schema = {
             "columns": [
-                {"name": "id", "type": "BIGINT", "nullable": False},
-                {"name": "optional_field", "type": "VARCHAR(50)", "nullable": True},
+                {"name": "id", "native_type": "BIGINT", "nullable": False},
+                {"name": "optional_field", "native_type": "VARCHAR(50)", "nullable": True},
             ]
         }
-        contract = DestinationSchemaContract(schema, type_mapper=type_mapper)
+        contract = SchemaContract(schema, type_mapper=type_mapper)
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply tests/unit/destination/test_schema_contract.py]: Resolved in f455643 — same fix as the later duplicate of this finding.

File: tests/unit/destination/test_schema_contract.py [hunk 17]
```diff
@@ -37,86 +39,163 @@ def type_mapper() -> TypeMapper:
...
-        assert len(result) == 1
-        assert result[0]["id"] == 1
-        assert result[0]["optional_field"] is None
+        assert result == [{"id": 1, "optional_field": None}]
+
+    def test_cast_arrow_batch_drops_extra_columns(self, type_mapper):
+        schema = {
+            "columns": [{"name": "id", "native_type": "BIGINT", "nullable": False}]
+        }
+        contract = SchemaContract(schema, type_mapper=type_mapper)
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply tests/unit/destination/test_schema_contract.py]: Resolved in f455643 — same fix as the later duplicate of this finding.

File: tests/unit/destination/test_schema_contract.py [hunk 18]
```diff
@@ -37,86 +39,163 @@ def type_mapper() -> TypeMapper:
...
+        """A non-nullable column absent from the incoming batch must fail
+        loudly. Silently null-filling would corrupt destinations that
+        enforce NOT NULL constraints at the database level."""
+        schema = {
+            "columns": [
+                {"name": "id", "native_type": "BIGINT", "nullable": False},
+                {"name": "amount", "native_type": "BIGINT", "nullable": False},
+            ]
+        }
+        contract = SchemaContract(schema, type_mapper=type_mapper)
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply tests/unit/destination/test_schema_contract.py]: Resolved in f455643 — same fix as the later duplicate of this finding.

File: tests/unit/destination/test_schema_contract.py [hunk 19]
```diff
@@ -37,86 +39,163 @@ def type_mapper() -> TypeMapper:
...
+
+    def test_cast_arrow_batch_unparseable_timestamp_raises(self, type_mapper):
+        """When every configured timestamp format fails, the cast must
+        raise instead of silently nulling the column."""
+        schema = {
+            "columns": [
+                {"name": "ts", "native_type": "TIMESTAMP", "nullable": True},
+            ]
+        }
+        contract = SchemaContract(schema, type_mapper=type_mapper)
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply tests/unit/destination/test_schema_contract.py]: Resolved in f455643 — same fix as the later duplicate of this finding.

File: tests/unit/destination/test_schema_contract.py [hunk 20]
```diff
@@ -37,86 +39,163 @@ def type_mapper() -> TypeMapper:
...
+
+        source = pa.RecordBatch.from_pylist([{"ts": "not-a-timestamp"}])
+        with pytest.raises(ValueError, match="cannot cast"):
+            contract.cast_arrow_batch(source)
+
+    def test_cast_arrow_batch_unparseable_date_raises(self, type_mapper):
+        schema = {
+            "columns": [{"name": "d", "native_type": "DATE", "nullable": True}]
+        }
+        contract = SchemaContract(schema, type_mapper=type_mapper)
```
[the remark this answers is not in the document]
Comment (kirill-andr) [reply tests/unit/destination/test_schema_contract.py]: Resolved in f455643 — same fix as the later duplicate of this finding.
